### ISSUE TITLE
Performance issues with large files

### ISSUE THREAD

**Issue by j-hui** [stevearc/aerial.nvim#37]
**Describe the bug**
When I open a massive .bib file (62k lines), aerial produces quite noticeable lag. In fact, the editor hangs completely for a couple of seconds at a time.

The lag goes away when I comment out aerial (and keep everything else in my setup).

Is there any way to improve the performance of aerial.nvim, or at least disable it beyond certain file sizes?

**System information**
- OS: linux
- Neovim version: 0.6.0
- AerialInfo:

```
Aerial Info
-----------
Filetype: bib
Configured backends:
  lsp (supported) (attached)
  treesitter (not supported)
  markdown (not supported)
Show symbols: Class, Constructor, Enum, Function, Interface, Method, Struct
```

**To Reproduce**
1. Load a large .bib file
2. Try to navigate around

Unfortunately, I cannot provide this .bib file since it belongs to my colleague.

**Additional context**

When Vim hangs, I noticed that sometimes when I mashed `<C-c>`, it would report that it was interrupted in the middle of Aerial's fold functions. While investigating that, I noticed that Aerial seems to conflict with FastFold when it comes to figuring out whether it should manage folds. I have `manage_folds` set to`'auto'`, and even when I explicitly set `foldmethod=syntax` in my `bib.vim`, Aerial still tries to take over, because it saw that FastFold had set the foldmethod to manual (I verified this by running `vim.cmd [[verbose set foldmethod]]` in `fold.lua`.

However, even with `manage_folds` disabled entirely, I'm still seeing really bad performance. But when I mash `<C-c>`, it no longer tells me that any plugin code was interrupted.

**Comment by stevearc** [comment]
I suspect that the performance issues are from ingesting and processing all of the symbols. I will admit that when I did my performance testing I didn't check files with anything close to 62k lines 😅

I took a profile and the time was dominated by the fold function, which I would expect. Due to how vim handles fold functions and how we store symbol data, it's an O(n<sup>2</sup>) operation. When n=62,000, that's not pretty. The immediate takeaway for me for this is that folding should be disabled by default. I think it's a neat feature, but I don't expect most people to use it so they shouldn't pay the performance cost for it.

I have to come back to this later, but I'm running another profile without folding. I expect that I will probably add an option to disable aerial on large files, and may also make some performance improvements if I can narrow down the bottlenecks.

**Comment by j-hui** [comment]
I also suspect that it might be the LSP server itself that's choking on the input size. I found that when I set the backend to treesitter, I get better performance. (I only got around to setting it up last night because I was curious if it would make a difference for performance, and it seems like it did.)

I'm not familiar with neovim's LSP architecture or how Aerial uses the LSP API, but do you know if Aerial talks to the LSP server asynchronously?

> I think it's a neat feature, but I don't expect most people to use it so they shouldn't pay the performance cost for it.

It is certainly useful for cases where LSP might provide better fold markers than syntax-driven heuristics (Treesitter or otherwise), but I'm not sure this is often the case. I'm still new to Treesitter but it seems like I can rely on it for folds when a decent parser is available; the set of filetypes for which that is true seems to be a superset of the filetypesfor which decent LSP servers are available.

I'm wondering if, rather than pre-computing all folds, Aerial provides any way to perform a manual fold for only the symbol the cursor is inside? Essentially doing `zf`/`:fold` rather than overloading `foldmethod`.

**Comment by stevearc** [comment]
The LSP server is almost certainly going to have issues, but that *shouldn't* affect the performance of neovim. All of the LSP communication is async, but single threaded. When we get all the symbols back from the server, those get processed in the main thread (which is what I suspect is causing the lag). I think I see some low-hanging fruit for performance optimizations, but it'll take a while to implement them all and there's no guarantee that it will *ever* be decent for a 62k line file, so in the meantime I added a `disable_max_lines` config option that defaults to 10,000. If you update, aerial should now be disabled on that file.

I'll keep this issue open to track the upcoming performance improvements.

**Comment by stevearc** [comment]
I've added some performance improvements, but I've been testing on a 400k line python file and there's still lag when opening aerial or modifying text. The only way I can think of to eliminate the lag is to chunk up the work and do it async to avoid blocking the main thread. This would be a significant amount of work and increase the complexity a lot. Given that the main use case for aerial is files with <10k lines, I'm going to say that for now the `disable_max_lines` option is good enough.

The good news is that we did get some notable perf wins! Super big files may still be laggy, but anything shy of that should be snappier now. The fold computation, in particular, is much faster.
